File: scripts/ble_messenger.py
#!/usr/bin/env python3

"""
Cross-platform BLE Messenger Script (Python Version)

This script provides a command-line interface for sending JSON messages
via Bluetooth Low Energy (BLE).

Usage:
    python ble_messenger.py scan                    # Scan for devices
    python ble_messenger.py connect <device-id>     # Connect to device
    python ble_messenger.py send <device-id> <message>  # Send message to device
    python ble_messenger.py broadcast <message>     # Send message to all connected devices
    python ble_messenger.py disconnect <device-id>  # Disconnect from device
    python ble_messenger.py list                    # List connected devices
    python ble_messenger.py help                    # Show help
"""

import asyncio
import json
import sys
import time
import uuid
from typing import Dict, List, Optional
import argparse

try:
    import asyncio
    from bleak import BleakClient, BleakScanner
    from bleak.backends.device import BLEDevice
except ImportError:
    print("Error: bleak library not found. Install it with: pip install bleak")
    sys.exit(1)


class BLEMessenger:
    def __init__(self):
        self.connected_devices: Dict[str, BleakClient] = {}
        self.is_scanning = False
        
        # Standard BLE service and characteristic UUIDs for messaging
        self.MESSAGING_SERVICE_UUID = "6E400001-B5A3-F393-E0A9-E50E24DCCA9E"  # Nordic UART Service
        self.TX_CHARACTERISTIC_UUID = "6E400002-B5A3-F393-E0A9-E50E24DCCA9E"  # TX Characteristic
        self.RX_CHARACTERISTIC_UUID = "6E400003-B5A3-F393-E0A9-E50E24DCCA9E"  # RX Characteristic

    async def scan_for_devices(self, duration: int = 10) -> List[BLEDevice]:
        """Scan for BLE devices that support the messaging service."""
        if self.is_scanning:
            print("Already scanning...")
            return []

        try:
            self.is_scanning = True
            print(f"Scanning for BLE devices for {duration} seconds...")
            
            devices = await BleakScanner.discover(
                timeout=duration,
                service_uuids=[self.MESSAGING_SERVICE_UUID]
            )
            
            print(f"Found {len(devices)} devices:")
            for device in devices:
                print(f"  - {device.name or 'Unknown'} ({device.address})")
                if device.metadata:
                    print(f"    RSSI: {device.metadata.get('rssi', 'N/A')}")
            
            return devices
            
        except Exception as e:
            print(f"Scan error: {e}")
            return []
        finally:
            self.is_scanning = False

    async def connect_to_device(self, device_id: str) -> bool:
        """Connect to a BLE device."""
        try:
            print(f"Connecting to device: {device_id}")
            
            client = BleakClient(device_id)
            await client.connect()
            
            # Discover services
            services = await client.get_services()
            print(f"Connected to: {device_id}")
            print(f"Services found: {len(services)}")
            
            self.connected_devices[device_id] = client
            return True
            
        except Exception as e:
            print(f"Connection error: {e}")
            return False

    async def disconnect_from_device(self, device_id: str) -> None:
        """Disconnect from a BLE device."""
        client = self.connected_devices.get(device_id)
        if client:
            try:
                await client.disconnect()
                del self.connected_devices[device_id]
                print(f"Disconnected from: {device_id}")
            except Exception as e:
                print(f"Disconnection error: {e}")
        else:
            print("Device not connected")

    async def send_message(self, device_id: str, message: str) -> bool:
        """Send a JSON message to a specific device."""
        client = self.connected_devices.get(device_id)
        if not client:
            print(f"Device not connected: {device_id}")
            return False

        try:
            ble_message = {
                "id": str(uuid.uuid4()),
                "timestamp": int(time.time() * 1000),
                "sender": "PythonScript",
                "content": message,
                "type": "text"
            }

            message_json = json.dumps(ble_message)
            message_bytes = message_json.encode('utf-8')

            # Split message into chunks if it's too large (BLE has MTU limits)
            chunk_size = 20
            chunks = [message_bytes[i:i + chunk_size] for i in range(0, len(message_bytes), chunk_size)]

            print(f"Sending message to {device_id}: {ble_message}")

            # Send each chunk
            for i, chunk in enumerate(chunks):
                await client.write_gatt_char(
                    self.TX_CHARACTERISTIC_UUID,
                    chunk
                )
                
                # Small delay between chunks
                if i < len(chunks) - 1:
                    await asyncio.sleep(0.05)

            print("Message sent successfully")
            return True
            
        except Exception as e:
            print(f"Send message error: {e}")
            return False

    async def broadcast_message(self, message: str) -> int:
        """Send a message to all connected devices."""
        if not self.connected_devices:
            print("No devices connected")
            return 0

        print(f"Broadcasting message to {len(self.connected_devices)} devices")
        
        tasks = [
            self.send_message(device_id, message)
            for device_id in self.connected_devices.keys()
        ]
        
        results = await asyncio.gather(*tasks, return_exceptions=True)
        success_count = sum(1 for result in results if result is True)
        
        print(f"Message sent to {success_count}/{len(self.connected_devices)} devices")
        return success_count

    def list_connected_devices(self) -> None:
        """List all connected devices."""
        if not self.connected_devices:
            print("No devices connected")
            return

        print("Connected devices:")
        for device_id in self.connected_devices.keys():
            print(f"  - {device_id}")

    async def cleanup(self) -> None:
        """Clean up all connections."""
        if self.connected_devices:
            print("Disconnecting from all devices...")
            tasks = [
                self.disconnect_from_device(device_id)
                for device_id in list(self.connected_devices.keys())
            ]
            await asyncio.gather(*tasks, return_exceptions=True)


async def main():
    """Main function to handle command-line interface."""
    parser = argparse.ArgumentParser(description="Cross-platform BLE Messenger")
    parser.add_argument("command", choices=[
        "scan", "connect", "send", "broadcast", "disconnect", "list", "help"
    ])
    parser.add_argument("args", nargs="*", help="Command arguments")
    
    args = parser.parse_args()

    if args.command == "help":
        show_help()
        return

    messenger = BLEMessenger()
    
    try:
        if args.command == "scan":
            await messenger.scan_for_devices()

        elif args.command == "connect":
            if len(args.args) < 1:
                print("Error: Please provide a device ID")
                return
            device_id = args.args[0]
            await messenger.connect_to_device(device_id)

        elif args.command == "disconnect":
            if len(args.args) < 1:
                print("Error: Please provide a device ID")
                return
            device_id = args.args[0]
            await messenger.disconnect_from_device(device_id)

        elif args.command == "send":
            if len(args.args) < 2:
                print("Error: Please provide device ID and message")
                return
            device_id = args.args[0]
            message = " ".join(args.args[1:])
            await messenger.send_message(device_id, message)

        elif args.command == "broadcast":
            if len(args.args) < 1:
                print("Error: Please provide a message")
                return
            message = " ".join(args.args)
            await messenger.broadcast_message(message)

        elif args.command == "list":
            messenger.list_connected_devices()

    except KeyboardInterrupt:
        print("\nInterrupted by user")
    except Exception as e:
        print(f"Error: {e}")
    finally:
        await messenger.cleanup()


def show_help():
    """Show help information."""
    help_text = """
Cross-platform BLE Messenger Script (Python Version)

Usage:
    python ble_messenger.py scan                    # Scan for devices
    python ble_messenger.py connect <device-id>     # Connect to device
    python ble_messenger.py send <device-id> <message>  # Send message to device
    python ble_messenger.py broadcast <message>     # Send message to all connected devices
    python ble_messenger.py disconnect <device-id>  # Disconnect from device
    python ble_messenger.py list                    # List connected devices
    python ble_messenger.py help                    # Show this help

Examples:
    python ble_messenger.py scan
    python ble_messenger.py connect AA:BB:CC:DD:EE:FF
    python ble_messenger.py send AA:BB:CC:DD:EE:FF "Hello, world!"
    python ble_messenger.py broadcast "Hello everyone!"
    python ble_messenger.py disconnect AA:BB:CC:DD:EE:FF
    python ble_messenger.py list

Requirements:
    pip install bleak
"""
    print(help_text)


if __name__ == "__main__":
    asyncio.run(main()) 

File: scripts/ble-messenger.js
#!/usr/bin/env node

/**
 * Cross-platform BLE Messenger Script
 * 
 * This script provides a command-line interface for sending JSON messages
 * via Bluetooth Low Energy (BLE).
 * 
 * Usage:
 *   node ble-messenger.js scan                    # Scan for devices
 *   node ble-messenger.js connect <device-id>     # Connect to device
 *   node ble-messenger.js send <message>          # Send message to connected device
 *   node ble-messenger.js broadcast <message>     # Send message to all connected devices
 *   node ble-messenger.js disconnect <device-id>  # Disconnect from device
 *   node ble-messenger.js list                    # List connected devices
 *   node ble-messenger.js help                    # Show help
 */

const { BleManager, State } = require('react-native-ble-plx');

class BLEScriptMessenger {
  constructor() {
    this.manager = new BleManager();
    this.connectedDevices = new Map();
    this.isScanning = false;
    
    // Standard BLE service and characteristic UUIDs for messaging
    this.MESSAGING_SERVICE = {
      serviceUUID: '6E400001-B5A3-F393-E0A9-E50E24DCCA9E', // Nordic UART Service
      characteristicUUID: '6E400002-B5A3-F393-E0A9-E50E24DCCA9E', // TX Characteristic
    };
  }

  async initialize() {
    try {
      const state = await this.manager.state();
      console.log('BLE State:', state);

      if (state === State.PoweredOn) {
        return true;
      }

      return new Promise((resolve) => {
        const subscription = this.manager.onStateChange((state) => {
          if (state === State.PoweredOn) {
            subscription.remove();
            resolve(true);
          } else if (state === State.PoweredOff) {
            subscription.remove();
            resolve(false);
          }
        }, true);
      });
    } catch (error) {
      console.error('BLE initialization error:', error);
      return false;
    }
  }

  async scanForDevices() {
    if (this.isScanning) {
      console.log('Already scanning...');
      return;
    }

    try {
      this.isScanning = true;
      console.log('Scanning for BLE devices...');

      await this.manager.startDeviceScan(
        [this.MESSAGING_SERVICE.serviceUUID],
        { allowDuplicates: false },
        (error, device) => {
          if (error) {
            console.error('Scan error:', error);
            this.isScanning = false;
            return;
          }

          if (device) {
            console.log(`Found device: ${device.name || 'Unknown'} (${device.id})`);
          }
        }
      );

      // Stop scanning after 10 seconds
      setTimeout(() => {
        this.stopScanning();
      }, 10000);

    } catch (error) {
      console.error('Start scanning error:', error);
      this.isScanning = false;
    }
  }

  stopScanning() {
    if (this.isScanning) {
      this.manager.stopDeviceScan();
      this.isScanning = false;
      console.log('Scan stopped');
    }
  }

  async connectToDevice(deviceId) {
    try {
      console.log(`Connecting to device: ${deviceId}`);
      
      const device = await this.manager.devices([deviceId]);
      if (device.length === 0) {
        console.error('Device not found');
        return false;
      }

      const connectedDevice = await device[0].connect();
      const discoveredDevice = await connectedDevice.discoverAllServicesAndCharacteristics();
      
      this.connectedDevices.set(deviceId, discoveredDevice);
      console.log(`Connected to: ${deviceId}`);
      
      return true;
    } catch (error) {
      console.error('Connection error:', error);
      return false;
    }
  }

  async disconnectFromDevice(deviceId) {
    const device = this.connectedDevices.get(deviceId);
    if (device) {
      try {
        await device.cancelConnection();
        this.connectedDevices.delete(deviceId);
        console.log(`Disconnected from: ${deviceId}`);
      } catch (error) {
        console.error('Disconnection error:', error);
      }
    } else {
      console.log('Device not connected');
    }
  }

  async sendMessage(deviceId, message) {
    const device = this.connectedDevices.get(deviceId);
    if (!device) {
      console.error('Device not connected:', deviceId);
      return false;
    }

    try {
      const bleMessage = {
        id: Math.random().toString(36).substr(2, 9),
        timestamp: Date.now(),
        sender: 'Script',
        content: message,
        type: 'text',
      };

      const messageJson = JSON.stringify(bleMessage);
      const messageBuffer = Buffer.from(messageJson, 'utf8');

      // Split message into chunks if it's too large
      const chunkSize = 20;
      const chunks = [];
      
      for (let i = 0; i < messageBuffer.length; i += chunkSize) {
        chunks.push(messageBuffer.slice(i, i + chunkSize));
      }

      // Send each chunk
      for (let i = 0; i < chunks.length; i++) {
        const chunk = chunks[i];
        await device.writeCharacteristicWithResponseForService(
          this.MESSAGING_SERVICE.serviceUUID,
          this.MESSAGING_SERVICE.characteristicUUID,
          chunk.toString('base64')
        );
        
        if (i < chunks.length - 1) {
          await new Promise(resolve => setTimeout(resolve, 50));
        }
      }

      console.log('Message sent successfully:', bleMessage);
      return true;
    } catch (error) {
      console.error('Send message error:', error);
      return false;
    }
  }

  async broadcastMessage(message) {
    if (this.connectedDevices.size === 0) {
      console.log('No devices connected');
      return 0;
    }

    const promises = Array.from(this.connectedDevices.keys()).map(deviceId =>
      this.sendMessage(deviceId, message)
    );

    const results = await Promise.all(promises);
    const successCount = results.filter(Boolean).length;
    console.log(`Message sent to ${successCount}/${this.connectedDevices.size} devices`);
    return successCount;
  }

  listConnectedDevices() {
    if (this.connectedDevices.size === 0) {
      console.log('No devices connected');
      return;
    }

    console.log('Connected devices:');
    this.connectedDevices.forEach((device, deviceId) => {
      console.log(`  - ${device.name || 'Unknown'} (${deviceId})`);
    });
  }

  destroy() {
    this.stopScanning();
    this.connectedDevices.forEach((device, deviceId) => {
      this.disconnectFromDevice(deviceId);
    });
    this.manager.destroy();
  }
}

// Command-line interface
async function main() {
  const args = process.argv.slice(2);
  const command = args[0];

  if (!command || command === 'help') {
    showHelp();
    return;
  }

  const messenger = new BLEScriptMessenger();
  
  try {
    const initialized = await messenger.initialize();
    if (!initialized) {
      console.error('Failed to initialize BLE. Please enable Bluetooth.');
      return;
    }

    switch (command) {
      case 'scan':
        await messenger.scanForDevices();
        break;

      case 'connect':
        const deviceId = args[1];
        if (!deviceId) {
          console.error('Please provide a device ID');
          return;
        }
        await messenger.connectToDevice(deviceId);
        break;

      case 'disconnect':
        const disconnectDeviceId = args[1];
        if (!disconnectDeviceId) {
          console.error('Please provide a device ID');
          return;
        }
        await messenger.disconnectFromDevice(disconnectDeviceId);
        break;

      case 'send':
        const targetDeviceId = args[1];
        const message = args[2];
        if (!targetDeviceId || !message) {
          console.error('Please provide device ID and message');
          return;
        }
        await messenger.sendMessage(targetDeviceId, message);
        break;

      case 'broadcast':
        const broadcastMessage = args[1];
        if (!broadcastMessage) {
          console.error('Please provide a message');
          return;
        }
        await messenger.broadcastMessage(broadcastMessage);
        break;

      case 'list':
        messenger.listConnectedDevices();
        break;

      default:
        console.error('Unknown command:', command);
        showHelp();
        break;
    }

  } catch (error) {
    console.error('Error:', error.message);
  } finally {
    // Keep the process alive for a bit to allow BLE operations to complete
    setTimeout(() => {
      messenger.destroy();
      process.exit(0);
    }, 2000);
  }
}

function showHelp() {
  console.log(`
Cross-platform BLE Messenger Script

Usage:
  node ble-messenger.js scan                    # Scan for devices
  node ble-messenger.js connect <device-id>     # Connect to device
  node ble-messenger.js send <device-id> <message>  # Send message to device
  node ble-messenger.js broadcast <message>     # Send message to all connected devices
  node ble-messenger.js disconnect <device-id>  # Disconnect from device
  node ble-messenger.js list                    # List connected devices
  node ble-messenger.js help                    # Show this help

Examples:
  node ble-messenger.js scan
  node ble-messenger.js connect ABC123
  node ble-messenger.js send ABC123 "Hello, world!"
  node ble-messenger.js broadcast "Hello everyone!"
  node ble-messenger.js disconnect ABC123
  node ble-messenger.js list
`);
}

// Handle process termination
process.on('SIGINT', () => {
  console.log('\nShutting down...');
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('\nShutting down...');
  process.exit(0);
});

// Run the script
if (require.main === module) {
  main().catch(console.error);
}

module.exports = BLEScriptMessenger; 